libraries/AP_Scripting/examples/plane_aerobatics.lua

```diff
@@ -73,6 +81,74 @@ function wrap_180(angle)
     return res
 end
 
+-- a PI controller implemented as a Lua object
+local function PI_controller(kP,kI,iMax)
+   -- the new instance. You can put public variables inside this self
+   -- declaration if you want to
+   local self = {}
+
+   -- private fields as locals
+   local _kP = kP or 0.0
+   local _kI = kI or 0.0
+   local _kD = kD or 0.0
+   local _iMax = iMax
+   local _last_t = nil
+   local _log_data = {}
+   local _I = 0
+   local _counter = 0
+
+   -- update the controller.
+   function self.update(target, current)
+      local now = millis():tofloat() * 0.001
+      if not _last_t then
+         _last_t = now
+      end
+      local dt = now - _last_t
+      _last_t = now
+      local err = target - current
+      _counter = _counter + 1
+
+      local P = _kP * err
+      _I = _I + _kI * err * dt
+      if _iMax then
+         _I = constrain(_I, -_iMax, iMax)
+      end
+      local I = _I
+      local ret = P + I
+
+      _log_data = { target, current, P, I, ret }
+      return ret
+   end
+
+   -- reset integrator to an initial value
+   function self.reset(integrator)
+      _I = integrator
+   end
+
+   function self.set_I(I)
+      _kI = I
+   end
+
+   function self.set_P(P)
+      _kP = P
+   end
+   
+   function self.set_Imax(Imax)
+      _iMax = Imax
+   end
+   
+   -- log the controller internals
+   function self.log(name)
+      logger.write(name,'Targ,Curr,P,I,Total','fffff',table.unpack(_log_data))
+   end
+
+   -- return the instance
+   return self
+end
+
+local height_PI = PI_controller(SCR_USER1:get(), SCR_USER2:get(), 20.0)
+local speed_PI = PI_controller(SCR_USER5:get(), SCR_USER6:get(), 100.0)
```

Comment:
Do we have to worry about windup for the speed controller?

Reply:
we should probably protect against saturation and stop wind-up, yes